firedrake/ml/pytorch.py

```diff
@@ -0,0 +1,212 @@
+ author = {Bouziani, Nacime and Ham, David A.},
+ booktitle = {{ICLR} 2023 {Workshop} on {Physics} for {Machine} {Learning}},
+ year = {2023},
+ doi = {10.48550/arXiv.2303.06871}
+}
+""")
+
+
+__all__ = ['FiredrakeTorchOperator', 'torch_operator', 'to_torch', 'from_torch']
+
+
+class FiredrakeTorchOperator(torch.autograd.Function):
+    """PyTorch custom operator representing a set of Firedrake operations expressed as a reduced functional `F`.
+
+    `FiredrakeTorchOperator` is a wrapper around :class:`torch.autograd.Function` that executes forward and backward
+    passes by directly calling the reduced functional `F`.
+
+    Parameters
+    ----------
+    metadata : dict
+               Dictionary used to stash Firedrake objects.
+    x_P : list of torch.Tensor
+          PyTorch tensors representing the inputs to the Firedrake operator `F`.
+
+    Returns
+    -------
+    torch.Tensor
+          PyTorch tensor representing the output of the Firedrake operator `F`.
+    """
+
+    def __init__(self):
+        super(FiredrakeTorchOperator, self).__init__()
+
+    # This method is wrapped by something cancelling annotation (probably 'with torch.no_grad()')
+    @staticmethod
+    def forward(ctx, metadata, *x_P):
+        """Forward pass of the PyTorch custom operator.
+        """
+        F = metadata['F']
+        V = metadata['V_controls']
+        # Convert PyTorch input (i.e. controls) to Firedrake
+        x_F = [from_torch(xi, Vi) for xi, Vi in zip(x_P, V)]
+        # Forward operator: delegated to pyadjoint.ReducedFunctional which recomputes the blocks on the tape
+        y_F = F(x_F)
+        # Stash metadata to the PyTorch context
+        ctx.metadata.update(metadata)
+        # Convert Firedrake output to PyTorch
+        y_P = to_torch(y_F)
+        return y_P.detach()
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        """Backward pass of the PyTorch custom operator.
+        """
+        F = ctx.metadata['F']
+        V = ctx.metadata['V_output']
+        # Convert PyTorch gradient to Firedrake
+        adj_input = from_torch(grad_output, V)
+        if isinstance(adj_input, Function):
+            adj_input = adj_input.vector()
+
+        # Compute adjoint model of `F`: delegated to pyadjoint.ReducedFunctional
+        adj_output = F.derivative(adj_input=adj_input)
+
+        # Tuplify adjoint output
+        adj_output = (adj_output,) if not isinstance(adj_output, collections.abc.Sequence) else adj_output
+
+        # None is for metadata arg in `forward`
+        return None, *[to_torch(di) for di in adj_output]
+
+
+def torch_operator(F):
+    """Cast a Firedrake reduced functional to a PyTorch operator.
+
+    The resulting :class:`~FiredrakeTorchOperator` will take PyTorch tensors as inputs and return PyTorch tensors as outputs.
+
+    Parameters
+    ----------
+    F : pyadjoint.ReducedFunctional
+        The reduced functional to wrap.
```

Comment:
It's rather restrictive to wrap only reduced functionals. Do you plan to extend this to arbitrary Firedrake programs?

Reply:
Despite the name, which is historic, ReducedFunctional can return any overloaded type. This means that this does enable wrapping of arbitrary Firedrake code. ReducedFunctional is simply the mechanism for expressing a Firedrake calculation as a function of controls, which is what you need if you're going to differentiate through it.

Reply:
I see. So it's possible to send a function that returns the result of `solve(F == 0, u, bcs=bcs)` to ReducedFunctional? I was confused with the test in this PR that uses a function named `solve_poisson` but returns `assemble(u ** 2 * dx)` instead of just returning `u` :
https://github.com/firedrakeproject/firedrake/blob/f05d119ac8302afb047a99d561e56584e4db60b4/tests/regression/test_pytorch_coupling.py#L86-L95

Reply:
Yes, you can. The output doesn't need to be scalar-valued as illustrated in the other test (`poisson_residual`):

https://github.com/firedrakeproject/firedrake/blob/f05d119ac8302afb047a99d561e56584e4db60b4/tests/regression/test_pytorch_coupling.py#L78-L82